### PR TITLE
create smaller unique files from boltdb shipper and other code refactorings

### PR DESCRIPTION
**Pull request by sandeepsukhani**
**What this PR does / why we need it**:
This PR does major overhaul of boltdb shipper code. Most of the functionality remains the same except it would shard index files by creating a new one every 15 mins.
The overhaul includes splitting upload and downloads code to a separate sub-package.
This PR still needs some final touches and tests.

**Checklist**
- [x] Tests updated

### CODE REVIEW

File: pkg/loki/modules.go [hunk 1]
```diff
@@ -34,7 +34,8 @@ import (
 	"github.com/grafana/loki/pkg/querier"
 	"github.com/grafana/loki/pkg/querier/queryrange"
 	loki_storage "github.com/grafana/loki/pkg/storage"
-	"github.com/grafana/loki/pkg/storage/stores/local"
+	"github.com/grafana/loki/pkg/storage/stores/shipper"
+	shipper_uploads "github.com/grafana/loki/pkg/storage/stores/shipper/uploads"
```
Comment (slim-bean) [inline pkg/loki/modules.go:38]: is there a collision that requires this? (I didn't look super close just aw it in the diff)

File: pkg/storage/stores/shipper/downloads/metrics.go [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+}
+
+func newMetrics(r prometheus.Registerer) *metrics {
+	m := &metrics{
+		filesDownloadDurationSeconds: &downloadTableDurationMetric{
+			periods: map[string]float64{},
+			gauge: promauto.With(r).NewGauge(prometheus.GaugeOpts{
+				Namespace: "loki_boltdb_shipper",
+				Name:      "initial_files_download_duration_seconds",
+				Help:      "Time (in seconds) spent in downloading of files per period, initially i.e for the first time",
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/downloads/metrics.go:68]: The help is confusing, I think you mean _for all periods_.

File: pkg/storage/stores/shipper/downloads/metrics.go [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+			})},
+		filesDownloadSizeBytes: &downloadTableBytesMetric{
+			periods: map[string]int64{},
+			gauge: promauto.With(r).NewGauge(prometheus.GaugeOpts{
+				Namespace: "loki_boltdb_shipper",
+				Name:      "initial_files_download_size_bytes",
+				Help:      "Size of files (in bytes) downloaded per period, initially i.e for the first time",
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/downloads/metrics.go:75]: Same idea this is for all periods.

File: pkg/storage/stores/shipper/downloads/table.go [hunk 4]
```diff
@@ -0,0 +1,400 @@
...
+	"github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"go.etcd.io/bbolt"
+)
+
+// timeout for downloading initial files for a table to avoid leaking resources by allowing it to take all the time.
+const downloadTimeout = 5 * time.Minute
+
+type BoltDBIndexClient interface {
+	QueryDB(ctx context.Context, db *bbolt.DB, query chunk.IndexQuery, callback func(chunk.IndexQuery, chunk.ReadBatch) (shouldContinue bool)) error
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/downloads/table.go:25]: I think you can use `chunk_util.Callback` here.

File: pkg/storage/stores/shipper/downloads/table.go [hunk 5]
```diff
@@ -0,0 +1,400 @@
...
+	boltdb *bbolt.DB
+}
+
+// Table is a collection of multiple files created for a same table by various ingesters.
+// All the public methods are concurrency safe and take care of mutexes to avoid any data race.
+type Table struct {
+	name              string
+	cacheLocation     string
+	metrics           *metrics
+	storageClient     chunk.ObjectClient
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/downloads/table.go:39]: You use only List and GetObject, you could lower the dependency requirement by creating your own interface declaring what you want to use.

Comment (cyriltovena) [reply pkg/storage/stores/shipper/downloads/table.go:39]: Like `BoltDBIndexClient` I also like the fact that it document the package.

File: pkg/storage/stores/shipper/downloads/table.go [hunk 6]
```diff
@@ -0,0 +1,400 @@
...
+	return
+}
+
+// Closes references to all the dbs.
+func (t *Table) Close() {
+	t.dbsMtx.Lock()
+	defer t.dbsMtx.Unlock()
+
+	// stop the initialization if it is still ongoing.
+	t.cancelFunc()
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/downloads/table.go:161]: The problem here is the lock will prevent you from actually cancelling.

I think you should use a different cancellation than the timeout, one that is initialized not the init goroutine, this way you don't race to cancel (that's why you need the lock) the first cancellation will also trigger the timeout context.

Comment (cyriltovena) [reply pkg/storage/stores/shipper/downloads/table.go:161]: See patch example:


```diff
diff --git a/pkg/storage/stores/shipper/downloads/table.go b/pkg/storage/stores/shipper/downloads/table.go
index 2107a589..99ee8603 100644
--- a/pkg/storage/stores/shipper/downloads/table.go
+++ b/pkg/storage/stores/shipper/downloads/table.go
@@ -49,6 +49,7 @@ type Table struct {
 }

 func NewTable(name, cacheLocation string, storageClient chunk.ObjectClient, boltDBIndexClient BoltDBIndexClient, metrics *metrics) *Table {
+	ctx, cancel := context.WithCancel(context.Background())
 	table := Table{
 		name:              name,
 		cacheLocation:     cacheLocation,
@@ -57,6 +58,7 @@ func NewTable(name, cacheLocation string, storageClient chunk.ObjectClient, bolt
 		boltDBIndexClient: boltDBIndexClient,
 		dbs:               map[string]*downloadedFile{},
 		ready:             make(chan struct{}),
+		cancelFunc:        cancel,
 	}

 	// keep the files collection locked until all the files are downloaded.
@@ -65,11 +67,9 @@ func NewTable(name, cacheLocation string, storageClient chunk.ObjectClient, bolt
 		defer table.dbsMtx.Unlock()
 		defer close(table.ready)

-		ctx, cancel := context.WithTimeout(context.Background(), downloadTimeout)
+		ctx, cancel := context.WithTimeout(ctx, downloadTimeout)
 		defer cancel()

-		table.cancelFunc = cancel
-
 		// Using background context to avoid cancellation of download when request times out.
 		// We would anyways need the files for serving next requests.
 		if err := table.init(ctx); err != nil {
@@ -154,12 +154,11 @@ func (t *Table) init(ctx context.Context) (err error) {

 // Closes references to all the dbs.
 func (t *Table) Close() {
-	t.dbsMtx.Lock()
-	defer t.dbsMtx.Unlock()
-
 	// stop the initialization if it is still ongoing.
 	t.cancelFunc()

+	t.dbsMtx.Lock()
+	defer t.dbsMtx.Unlock()
 	for name, db := range t.dbs {
 		if err := db.boltdb.Close(); err != nil {
 			level.Error(util.Logger).Log("msg", fmt.Errorf("failed to close file %s for table %s", name, t.name))

```

File: pkg/storage/stores/shipper/downloads/table.go (hunk 4 above)
Comment (sandeepsukhani) [reply pkg/storage/stores/shipper/downloads/table.go:25]: I had tried it. The problem is the function in boltdb index client in Cortex defines function using `callback func(chunk.IndexQuery, chunk.ReadBatch) (shouldContinue bool)` instead of `chunk_util.Callback`. If I change it here without changing in Cortex, interface is not satisfied anymore. We will have to change it in Cortex first.

**Comment by sandeepsukhani** [comment]
@cyriltovena Thanks for the feedback! I have addressed them all.
Please feel free to point out any problems that you see, if any.

File: pkg/storage/stores/shipper/downloads/table.go (hunk 4 above)
Comment (cyriltovena) [reply pkg/storage/stores/shipper/downloads/table.go:25]: ho then just define it here.

Comment (sandeepsukhani) [reply pkg/storage/stores/shipper/downloads/table.go:25]: Sorry, didn't get you. Don't we have to change it in Cortex first?

File: pkg/storage/stores/shipper/downloads/table_manager.go [hunk 7]
```diff
@@ -0,0 +1,166 @@
...
+		cfg:             cfg,
+		boltIndexClient: boltIndexClient,
+		storageClient:   storageClient,
+		tables:          make(map[string]*Table),
+		metrics:         newMetrics(registerer),
+		done:            make(chan struct{}),
+	}, nil
+}
+
+func (tm *TableManager) loop() {
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/downloads/table_manager.go:48]: loop is unused ? doesn't seems good.

File: pkg/storage/stores/shipper/downloads/table_manager.go [hunk 8]
```diff
@@ -0,0 +1,166 @@
...
+
+func (tm *TableManager) QueryPages(ctx context.Context, queries []chunk.IndexQuery, callback chunk_util.Callback) error {
+	return chunk_util.DoParallelQueries(ctx, tm.query, queries, callback)
+}
+
+func (tm *TableManager) query(ctx context.Context, query chunk.IndexQuery, callback chunk_util.Callback) error {
+	table := tm.getOrCreateTable(query.TableName)
+
+	// let us check if table is ready for use while also honoring the context timeout
+	select {
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/downloads/table_manager.go:95]: 👍

File: pkg/storage/stores/shipper/downloads/table_manager.go [hunk 9]
```diff
@@ -0,0 +1,166 @@
...
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-table.IsReady():
+	}
+
+	if table.Err() != nil {
+		tm.tablesMtx.Lock()
+		defer tm.tablesMtx.Unlock()
+
+		delete(tm.tables, query.TableName)
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/downloads/table_manager.go:105]: I think you should log table deletion like you're doing for addition. This would help debug if anything goes wrong.

File: pkg/storage/stores/shipper/downloads/table.go [hunk 10]
```diff
@@ -0,0 +1,406 @@
...
+	t.dbs = map[string]*downloadedFile{}
+}
+
+// Queries all the dbs for index.
+func (t *Table) Query(ctx context.Context, query chunk.IndexQuery, callback chunk_util.Callback) error {
+	if t.err != nil {
+		return t.err
+	}
+
+	t.dbsMtx.RLock()
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/downloads/table.go:184]: I think the code for waiting initialization should be before this. I know it is the case already but it should included in this function and not in the table manager, because If I use table without tablemanager I can get deadlocked here.

Another option is to make Table not exported in this package.

File: pkg/storage/stores/shipper/downloads/table_manager.go (hunk 9 above)
Comment (cyriltovena) [reply pkg/storage/stores/shipper/downloads/table_manager.go:111]: with the error that triggered the deletion may be ?

File: pkg/storage/stores/shipper/downloads/table_manager.go [hunk 11]
```diff
@@ -0,0 +1,172 @@
...
+	chunk_util "github.com/cortexproject/cortex/pkg/chunk/util"
+	pkg_util "github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const cacheCleanupInterval = 24 * time.Hour
+
+type Config struct {
+	CacheDir     string
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/downloads/table_manager.go:19]: Should we support flags too ? Usually nice for documenting fields or overriding. Not sure @slim-bean ?

File: pkg/storage/stores/shipper/downloads/table_manager.go [hunk 12]
```diff
@@ -0,0 +1,172 @@
...
+	"time"
+
+	"github.com/cortexproject/cortex/pkg/chunk"
+	chunk_util "github.com/cortexproject/cortex/pkg/chunk/util"
+	pkg_util "github.com/cortexproject/cortex/pkg/util"
+	"github.com/go-kit/kit/log/level"
+	"github.com/prometheus/client_golang/prometheus"
+)
+
+const cacheCleanupInterval = 24 * time.Hour
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/downloads/table_manager.go:16]: 24h that's seems relatively big. Specially that you're checking the time last used with TTL config, worst case it's a noop. Let's pick something lower.

File: pkg/storage/stores/shipper/downloads/table_manager.go [hunk 13]
```diff
@@ -0,0 +1,172 @@
...
+	}
+
+	return nil
+}
+
+func (tm *TableManager) cleanupCache() error {
+	tm.tablesMtx.Lock()
+	defer tm.tablesMtx.Unlock()
+
+	for name, table := range tm.tables {
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/downloads/table_manager.go:159]: You should test this function.

File: pkg/storage/stores/shipper/downloads/table.go (hunk 10 above)
Comment (cyriltovena) [reply pkg/storage/stores/shipper/downloads/table.go:184]: May be easier to just rename Table to table.

File: pkg/storage/stores/shipper/downloads/table.go [hunk 14]
```diff
@@ -0,0 +1,406 @@
...
+			level.Error(util.Logger).Log("msg", fmt.Errorf("failed to close file %s for table %s", name, t.name))
+		}
+	}
+
+	t.dbs = map[string]*downloadedFile{}
+}
+
+// Queries all the dbs for index.
+func (t *Table) Query(ctx context.Context, query chunk.IndexQuery, callback chunk_util.Callback) error {
+	if t.err != nil {
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/downloads/table.go:180]: I would move this after the lock.

File: pkg/storage/stores/shipper/downloads/table_manager.go (hunk 11 above)
Comment (cyriltovena) [reply pkg/storage/stores/shipper/downloads/table_manager.go:19]: Ho just realized this is passed down another configuration that has flags support. Nevermind.

File: pkg/storage/stores/shipper/uploads/table.go [hunk 15]
```diff
@@ -0,0 +1,325 @@
...
+	db, ok := lt.dbs[shard]
+	if !ok {
+		lt.dbsMtx.RUnlock()
+
+		err := lt.addDB(shard)
+		if err != nil {
+			return err
+		}
+
+		lt.dbsMtx.RLock()
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/uploads/table.go:146]: if addDb returns the added db you would not have to do a lock here.

File: pkg/storage/stores/shipper/uploads/table_manager.go [hunk 16]
```diff
@@ -0,0 +1,230 @@
...
+func (tm *TableManager) loop() {
+	defer tm.wg.Done()
+
+	syncTicker := time.NewTicker(tm.cfg.UploadInterval)
+	defer syncTicker.Stop()
+
+	for {
+		select {
+		case <-syncTicker.C:
+			err := tm.uploadTables(context.Background())
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/uploads/table_manager.go:67]: I don't think you should be using a context.Background here.

Comment (cyriltovena) [reply pkg/storage/stores/shipper/uploads/table_manager.go:67]: What is the issue when you cancel this ?

File: pkg/storage/stores/shipper/uploads/table_manager.go [hunk 17]
```diff
@@ -0,0 +1,230 @@
...
+			return
+		}
+	}
+}
+
+func (tm *TableManager) Stop() {
+	close(tm.done)
+	tm.wg.Wait()
+
+	err := tm.uploadTables(context.Background())
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/uploads/table_manager.go:81]: Same here. plus both of  `uploadTables` references use it only with Background.

File: pkg/storage/stores/shipper/uploads/table.go [hunk 18]
```diff
@@ -0,0 +1,321 @@
...
+	}, nil
+}
+
+// Query serves the index by querying all the open dbs.
+func (lt *Table) Query(ctx context.Context, query chunk.IndexQuery, callback chunk_util.Callback) error {
+	lt.dbsMtx.RLock()
+	defer lt.dbsMtx.RUnlock()
+
+	for _, db := range lt.dbs {
+		if err := lt.boltdbIndexClient.QueryDB(ctx, db, query, callback); err != nil {
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/uploads/table.go:98]: I wonder if this shouldn't be run on multiple goroutine ? Technically reading a file is not CPU bound so you should gain something by not waiting an operation and fire more at once.

Comment (cyriltovena) [reply pkg/storage/stores/shipper/uploads/table.go:98]: I'm not sure but a benchmark could help up figure this out.

File: pkg/storage/stores/shipper/uploads/table.go [hunk 19]
```diff
@@ -0,0 +1,321 @@
...
+
+		lt.uploadedDBsMtimeMtx.RLock()
+		uploadedDBMtime, ok := lt.uploadedDBsMtime[name]
+		lt.uploadedDBsMtimeMtx.RUnlock()
+
+		if ok && !uploadedDBMtime.Before(stat.ModTime()) {
+			continue
+		}
+
+		err = lt.uploadDB(ctx, name, db)
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/uploads/table.go:202]: Same I think we should upload multiple DB at once. Not easy to write so we can start with this if you prefer. But you're less at risk of corrupted data if you run faster :)

File: pkg/storage/stores/shipper/uploads/table_manager.go [hunk 20]
```diff
@@ -0,0 +1,230 @@
...
+		return errors.New("invalid write batch")
+	}
+
+	for tableName, tableWrites := range boltWriteBatch.Writes {
+		table, err := tm.getOrCreateTable(tableName)
+		if err != nil {
+			return err
+		}
+
+		err = table.Write(ctx, tableWrites)
```
Comment (cyriltovena) [inline pkg/storage/stores/shipper/uploads/table_manager.go:120]: Again this is just a question should this be done in parallel ?

File: pkg/storage/stores/shipper/downloads/table_manager.go (hunk 7 above)
Comment (sandeepsukhani) [reply pkg/storage/stores/shipper/downloads/table_manager.go:53]: Sorry, my bad. Thanks for taking care of it!

File: pkg/storage/stores/shipper/downloads/table.go (hunk 10 above)
Comment (sandeepsukhani) [reply pkg/storage/stores/shipper/downloads/table.go:184]: I have made public methods to give a guarantee that usage of them is concurrency safe. The `Table` is not meant to be used without the table manager. Having public methods(for concurrency guarantees) on a private type would look weird. It would be better to move readiness check here from table manager given we want to support using `Table` without table manager.
What do you think?

File: pkg/storage/stores/shipper/downloads/table_manager.go (hunk 12 above)
Comment (sandeepsukhani) [reply pkg/storage/stores/shipper/downloads/table_manager.go:16]: How about 12h?

File: pkg/storage/stores/shipper/uploads/table_manager.go (hunk 16 above)
Comment (sandeepsukhani) [reply pkg/storage/stores/shipper/uploads/table_manager.go:72]: I don't think we would ever want to cancel that context, not even while stopping to make sure we stop the service only after uploading the files. We could have a context with a minimum of 10 mins timeout to avoid waiting on it forever. What do you think? Is that what you are looking for?

File: pkg/storage/stores/shipper/uploads/table_manager.go (hunk 17 above)
Comment (sandeepsukhani) [reply pkg/storage/stores/shipper/uploads/table_manager.go:86]: Same here, we can have a 10 mins timeout on the context.

File: pkg/storage/stores/shipper/uploads/table.go (hunk 18 above)
Comment (sandeepsukhani) [reply pkg/storage/stores/shipper/uploads/table.go:98]: I think let us have a basic version out, we can do all these optimizations when we see a problem there. We already parallelize queries for different tables, we can do the same here with bounded concurrency. What do you think?

File: pkg/storage/stores/shipper/uploads/table.go (hunk 19 above)
Comment (sandeepsukhani) [reply pkg/storage/stores/shipper/uploads/table.go:202]: Same as above, let us do the optimizations when we need them because they add complexity and would also, in turn, increase the size of the PR.

File: pkg/storage/stores/shipper/uploads/table_manager.go (hunk 20 above)
Comment (sandeepsukhani) [reply pkg/storage/stores/shipper/uploads/table_manager.go:120]: Same as above.

File: pkg/storage/stores/shipper/downloads/table_manager.go (hunk 9 above)
Comment (sandeepsukhani) [reply pkg/storage/stores/shipper/downloads/table_manager.go:113]: Added the logs. Thanks!

File: pkg/storage/stores/shipper/downloads/table.go (hunk 10 above)
Comment (cyriltovena) [reply pkg/storage/stores/shipper/downloads/table.go]: Either you make `Table` => `table` or you move the readiness check here. Up to you !

File: pkg/storage/stores/shipper/downloads/table_manager.go (hunk 12 above)
Comment (cyriltovena) [reply pkg/storage/stores/shipper/downloads/table_manager.go]: yeah one hour is probably better.

File: pkg/storage/stores/shipper/uploads/table.go (hunk 18 above)
Comment (cyriltovena) [reply pkg/storage/stores/shipper/uploads/table.go:98]: Yep no problem !

File: pkg/storage/stores/shipper/uploads/table.go (hunk 19 above)
Comment (cyriltovena) [reply pkg/storage/stores/shipper/uploads/table.go:204]: Sure !

File: pkg/storage/stores/shipper/uploads/table_manager.go (hunk 17 above)
Comment (cyriltovena) [reply pkg/storage/stores/shipper/uploads/table_manager.go:89]: replied above

File: pkg/storage/stores/shipper/downloads/table_manager.go (hunk 12 above)
Comment (sandeepsukhani) [reply pkg/storage/stores/shipper/downloads/table_manager.go]: I have pushed the code with interval set to 1h already. Resolving this.

File: pkg/storage/stores/shipper/uploads/table_manager.go (hunk 16 above)
Comment (cyriltovena) [reply pkg/storage/stores/shipper/uploads/table_manager.go:73]: My problem is mostly when using context.Background you're blocking the waitgroup and you'll likely be killed and you don't know when. I'm wondering if this could be a problem ? Basically being stopped before gracefully stopping.

First I think you should cancel the loop because the upload happen again anyways after stop. Am I wrong ?

Second 10min would get just shutdown by Kubernetes see https://cloud.google.com/blog/products/gcp/kubernetes-best-practices-terminating-with-grace

> At this point, Kubernetes waits for a specified time called the termination grace period. By default, this is 30 seconds. It’s important to note that this happens in parallel to the preStop hook and the SIGTERM signal. Kubernetes does not wait for the preStop hook to finish.

Basically you're going to get SIGKILL after 30s or more depending on what we have configured, for sure not 10min.

Let's change the Close function to give a 1m timeout, and log if it has been cancelled or success, so we can see if we get killed before finishing.

File: pkg/storage/stores/shipper/uploads/table_manager.go (hunk 17 above)
Comment (sandeepsukhani) [reply pkg/storage/stores/shipper/uploads/table_manager.go:89]: Can you please check if you missed posting the comment? I can't see it yet.

File: pkg/storage/stores/shipper/downloads/table_manager.go (hunk 13 above)
Comment (sandeepsukhani) [reply pkg/storage/stores/shipper/downloads/table_manager.go:165]: I have added the tests for this. Thanks!

File: pkg/storage/stores/shipper/uploads/table_manager.go (hunk 16 above)
Comment (sandeepsukhani) [reply pkg/storage/stores/shipper/uploads/table_manager.go:73]: If I am not wrong our ingesters don't get killed within 30s of receiving SIGTERM because we do chunk transfers and it could take a couple of minutes. I checked the config and it seems we configure termination grace period to 80 minutes, see https://github.com/grafana/loki/blob/a0dc10a03a0a82fef7e4783ceeac0fecd65a0c58/production/ksonnet/loki/ingester.libsonnet#L31

File: pkg/storage/stores/shipper/downloads/table.go (hunk 10 above)
Comment (sandeepsukhani) [reply pkg/storage/stores/shipper/downloads/table.go]: I have pushed the code. Thanks!

File: pkg/storage/stores/shipper/uploads/table_manager.go (hunk 16 above)
Comment (slim-bean) [reply pkg/storage/stores/shipper/uploads/table_manager.go:73]: I think this context should have a timeout and should be configured to a value less than the termination grace period and we should have explicit info level logging on when a table upload starts and completed.

I don't like the idea of hiding a table upload into the background without any visibility, many people run Loki with much more aggressive shutdown requirements and it shoudl be clear from the logs if table uploads succeeded or if the process was killed before they completed, have a timeout less than the grace period would ensure we log that upload filed before we gave up and shutdown.

Comment (slim-bean) [reply pkg/storage/stores/shipper/uploads/table_manager.go:73]: After some more discussion with @sandeepsukhani , we already try to flush chunks forever on shutdown it makes sense to try to upload tables forever on shutdown.

I would like to see the explicit logging of both started and succeeded (as well as error) so that someone can determine from their logs if all tables were uploaded.
